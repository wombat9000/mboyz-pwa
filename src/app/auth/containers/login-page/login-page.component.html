<div class="padded">
  <h2>Hallo!</h2>
  <p>Du musst Dich bei Facebook anmelden, um den Urlaubsplaner zu nutzen.</p>
</div>

<div *ngIf="pending$ | async">
  <mat-spinner></mat-spinner>
</div>
<button mat-button class="button-fb" (click)="fbLogin()">Mit Facebook verbinden</button>

<div class="login-error padded" *ngIf="error$ | async as error">
  <h3>Es ist ein Fehler aufgetreten.</h3>
  <p>{{error}}</p>
</div>
