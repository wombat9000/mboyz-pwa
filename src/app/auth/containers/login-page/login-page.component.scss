@import '~@angular/material/theming';
@import '../../../../mboyz-theme.scss';

:host {

  .padded {
    padding: 5px;
  }

  .button-fb {
    width: 100%;
    background-color: $fb-blue;
    color: $fb-white;
  }

  mat-spinner {
    margin: 50% auto;
  }
}
