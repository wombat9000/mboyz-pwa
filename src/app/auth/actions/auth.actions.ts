import {Action} from '@ngrx/store';
import {User} from '../services/auth.service';

export enum AuthActionTypes {
  AUTHORISE = '[Auth] Authorise',
  UNAUTHORISED = '[Auth] Unauthorised',
  FACEBOOK_LOGIN = '[Auth] FacebookLogin',
  LOGOUT = '[Auth] Logout',
  NOT_AUTHENTICATED = '[Auth] Not Authenticated',
  LOGIN_SUCCESS = '[Auth] Login Success',
  LOGIN_FAILURE = '[Auth] Login Failure',
}

export class Unauthorised implements Action {
  readonly type = AuthActionTypes.UNAUTHORISED;
  constructor(public payload: { url: string }) {}
}

export class Authorise implements Action {
  readonly type = AuthActionTypes.AUTHORISE;
  constructor(public payload: { user: User, url: string }) {}
}

export class FacebookLogin implements Action {
  readonly type = AuthActionTypes.FACEBOOK_LOGIN;
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.LOGIN_SUCCESS;
  constructor(public payload: { user: User }) {}
}

export class NotAuthenticated implements Action {
  readonly type = AuthActionTypes.NOT_AUTHENTICATED;
}

export class LoginFailure implements Action {
  readonly type = AuthActionTypes.LOGIN_FAILURE;
  constructor(public payload: { error: string }) {}
}

export class Logout implements Action {
  readonly type = AuthActionTypes.LOGOUT;
}

export type AuthActions =
  | Unauthorised
  | Authorise
  | FacebookLogin
  | LoginSuccess
  | LoginFailure
  | Logout
  | NotAuthenticated;
