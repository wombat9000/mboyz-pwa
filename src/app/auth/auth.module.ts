import {ModuleWithProviders, NgModule} from '@angular/core';
import {UserFirestore} from './services/user-firestore.service';
import {AuthService} from './services/auth.service';
import {AuthGuard} from './services/auth.guard';
import {MaterialModule} from '../material';
import {StoreModule} from '@ngrx/store';
import {reducers} from './reducers';
import {EffectsModule} from '@ngrx/effects';
import {AuthEffects} from './effects/auth.effects';
import {LoginPageComponent} from './containers/login-page/login-page.component';
import {RouterModule} from '@angular/router';
import {CommonModule} from '@angular/common';

@NgModule({
  imports: [CommonModule, MaterialModule],
  providers: [UserFirestore, AuthService, AuthGuard],
  declarations: [LoginPageComponent],
  exports: [LoginPageComponent]
})
export class AuthModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: RootAuthModule,
      providers: [AuthService, AuthGuard],
    };
  }
}

@NgModule({
  imports: [
    AuthModule,
    RouterModule.forChild([{path: 'login', component: LoginPageComponent}]),
    StoreModule.forFeature('auth', reducers),
    EffectsModule.forFeature([AuthEffects]),
  ],
})
export class RootAuthModule {
}
