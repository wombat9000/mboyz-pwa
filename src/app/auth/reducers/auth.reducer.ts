import {MtravelUser} from '../services/auth.service';
import {AuthActions, AuthActionTypes} from '../actions/auth.actions';

export interface State {
  loggedIn: boolean;
  user: MtravelUser | null;
}

export const initialState: State = {
  loggedIn: false,
  user: null,
};

export function reducer(state = initialState, action: AuthActions): State {
  switch (action.type) {
    case AuthActionTypes.AUTHORISE:
    case AuthActionTypes.LOGIN_SUCCESS: {
      return {
        loggedIn: true,
        user: action.payload.user,
      };
    }

    case AuthActionTypes.NOT_AUTHENTICATED: {
      return initialState;
    }

    default: {
      return state;
    }
  }
}

export const getUser = (state: State) => state.user;
