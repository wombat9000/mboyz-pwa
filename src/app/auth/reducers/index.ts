import * as fromAuth from './auth.reducer';
import * as fromLoginPage from './login-page.reducer';
import {ActionReducerMap, createFeatureSelector, createSelector} from '@ngrx/store';


export interface AuthState {
  status: fromAuth.State;
  loginPage: fromLoginPage.State;
}

export interface State {
  auth: AuthState;
}

export const reducers: ActionReducerMap<any> = {
  status: fromAuth.reducer,
  loginPage: fromLoginPage.reducer,
};

export const selectAuthState = createFeatureSelector<AuthState>('auth');

export const selectAuthStatusState = createSelector(
  selectAuthState,
  (state: AuthState) => state.status
);

export const selectLoginPageState = createSelector(
  selectAuthState,
  (state: AuthState) => state.loginPage
);

export const getUser = createSelector(
  selectAuthStatusState,
  fromAuth.getUser
);

export const getLoginPagePending = createSelector(
  selectLoginPageState,
  fromLoginPage.getPending
);

export const getErrorMessage = createSelector(
  selectLoginPageState,
  fromLoginPage.getError
);
