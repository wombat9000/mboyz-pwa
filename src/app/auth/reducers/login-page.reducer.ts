import {AuthActions, AuthActionTypes} from '../actions/auth.actions';

export interface State {
  error?: string;
  pending: boolean;
}

export const initialState: State = {
  pending: false
};

export function reducer(state = initialState, action: AuthActions): State {
  switch (action.type) {
    case AuthActionTypes.GET_USER:
    case AuthActionTypes.FB_LOGIN: {
      return {pending: true};
    }
    case AuthActionTypes.LOGIN_SUCCESS: {
      return initialState;
    }
    case AuthActionTypes.LOGIN_FAILURE: {
      return {pending: false, error: action.payload.error};
    }
    default: {
      return state;
    }
  }
}

export const getPending = (state: State) => state.pending;
export const getError = (state: State) => state.error;

