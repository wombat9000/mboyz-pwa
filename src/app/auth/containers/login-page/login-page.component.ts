import {Component} from '@angular/core';
import {Store} from '@ngrx/store';

import {FbLogin} from '../../actions/auth.actions';
import * as fromAuth from '../../reducers';
import {Observable} from 'rxjs/Observable';


@Component({
  selector: 'app-login-page',
  templateUrl: './login-page.component.html',
  styleUrls: ['./login-page.component.scss']
})
export class LoginPageComponent {
  pending$: Observable<boolean> = this.store.select(fromAuth.getLoginPagePending);
  error$: Observable<string> = this.store.select(fromAuth.getErrorMessage);

  constructor(private store: Store<fromAuth.State>) {
  }

  fbLogin() {
    this.store.dispatch(new FbLogin());
  }
}
