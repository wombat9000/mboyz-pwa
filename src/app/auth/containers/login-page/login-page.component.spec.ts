import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {By} from '@angular/platform-browser';
import {click} from '../../../test-support/functions';
import * as fromAuth from '../../reducers';
import {combineReducers, Store, StoreModule} from '@ngrx/store';
import {FacebookLogin, LoginFailure} from '../../actions/auth.actions';
import {LoginPageComponent} from './login-page.component';
import {CUSTOM_ELEMENTS_SCHEMA} from '@angular/core';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';


class LoginPO {
  constructor(private fixture: ComponentFixture<LoginPageComponent>) {
    this.fixture.detectChanges();
  }

  spinner() {
    return this.fixture.debugElement.query(By.css('mat-spinner'));
  }

  errorInfo() {
    return this.fixture.debugElement.query(By.css('.login-error'));
  }

  clickFBLogin() {
    click(this.fixture.debugElement.query(By.css('.button-fb')));
  }
}

describe('Login Page', () => {
  let store: Store<fromAuth.State>;
  let fixture: ComponentFixture<LoginPageComponent>;
  let loginPO: LoginPO;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        StoreModule.forRoot({
          auth: combineReducers(fromAuth.reducers),
        })
      ],
      declarations: [LoginPageComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
      .compileComponents();
  }));


  beforeEach(async () => {
    store = TestBed.get(Store);

    fixture = TestBed.createComponent(LoginPageComponent);
    loginPO = new LoginPO(fixture);
    await fixture.whenStable();
  });

  describe('initial state', () => {
    it('should trigger facebook login on click', () => {
      spyOn(store, 'dispatch');

      loginPO.clickFBLogin();

      expect(store.dispatch).toHaveBeenCalledWith(new FacebookLogin());
    });

    it('should not show spinner', () => {
      fixture.detectChanges();
      expect(loginPO.spinner()).toBeNull();
    });

    it('should should not show error', () => {
      fixture.detectChanges();
      expect(loginPO.errorInfo()).toBeNull();
    });
  });

  describe('login pending', () => {
    beforeEach(() => {
      store.dispatch(new FacebookLogin());
    });

    it('should show spinner', () => {
      fixture.detectChanges();

      expect(loginPO.spinner()).toBeTruthy();
    });
  });

  describe('login failure', () => {
    beforeEach(() => {
      store.dispatch(new LoginFailure({error: 'someError'}));
    });

    it('should show error message', () => {
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });
});
