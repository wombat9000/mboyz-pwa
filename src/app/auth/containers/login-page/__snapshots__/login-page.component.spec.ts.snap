// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page login failure should show error message 1`] = `
<app-login-page
  error$={[Function Store]}
  pending$={[Function Store]}
  store={[Function Store]}
>
  <div
    class="padded"
  >
    <h2>
      Hallo!
    </h2>
    <p>
      Du musst Dich bei Facebook anmelden, um den Urlaubsplaner zu nutzen.
    </p>
  </div><button
    class="button-fb"
    mat-button=""
  >
    Mit Facebook verbinden
  </button><div
    class="login-error padded"
  >
    <h3>
      Es ist ein Fehler aufgetreten.
    </h3>
    <p>
      someError
    </p>
  </div>
</app-login-page>
`;
